### PR TITLE
[ADR] ProposeTx

### PR DESCRIPTION
**Pull request by iantanwx**
This adds an ADR entry addressing the implementation of a `ProposeTx`
method in the ABCI proposed in #1776. Fundamentally, this proposal gives
some control of block proposals to the application. The initial use case is
to support the Minimal Viable Plasma specification.

cc @ebuchman

### CODE REVIEW

File: docs/architecture/adr-012-ABCI-propose-tx.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+message ResponseProposeTx {
+  int64 num_tx = 1; // the number of tx to include in proposed block
+  repeated bytes txs = 2; // ordered transaction data to include in block
+  bool exclusive = 3; // whether the block should include other transactions (from `mempool`)
+}
+```
+
+`ProposeTx` would be called by before `mempool.Reap` at this
+[line](https://github.com/tendermint/tendermint/blob/master/consensus/state.go#L906).
```
Comment (milosevic) [inline docs/architecture/adr-012-ABCI-propose-tx.md]: Shoudn't it be before this line: https://github.com/tendermint/tendermint/blob/master/consensus/state.go#L928?

File: docs/architecture/adr-012-ABCI-propose-tx.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+```
+ProposeTx(RequestProposeTx) ResponseProposeTx
+```
+
+Where `RequestProposeTx` and `ResponseProposeTx` are `message`s with the
+following shapes:
+
+```
+message RequestProposeTx {
+  int64 next_block_height = 1; // height of the block the proposed tx would be part of
```
Comment (milosevic) [inline docs/architecture/adr-012-ABCI-propose-tx.md]: Why we need RequestProposeTx? Do you expect app logic to differ based on the current height and proposer?

File: docs/architecture/adr-012-ABCI-propose-tx.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  Validator proposer = 2; // the proposer details
+}
```
Comment (milosevic) [inline docs/architecture/adr-012-ABCI-propose-tx.md]: Maybe Tendermint can constrain number of txs (maximum size) an app can provide as part of this ProposeTx logic? This can then be configured/governed as part of ConsensusParams.

File: docs/architecture/adr-012-ABCI-propose-tx.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+message ResponseProposeTx {
+  int64 num_tx = 1; // the number of tx to include in proposed block
+  repeated bytes txs = 2; // ordered transaction data to include in block
+  bool exclusive = 3; // whether the block should include other transactions (from `mempool`)
+}
+```
+
+`ProposeTx` would be called by before `mempool.Reap` at this
+[line](https://github.com/tendermint/tendermint/blob/master/consensus/state.go#L906).
+Depending on whether `exclusive` is `true` or `false`, the proposed
```
Comment (milosevic) [inline docs/architecture/adr-012-ABCI-propose-tx.md]: Maybe say instead: "If `exclusive` is `true`, the proposal block will not contain other transactions from the mempool; otherwise, the proposed transactions will be included in the proposal block ordered before the transactions obtained by `mempool.Reap`.

File: docs/architecture/adr-012-ABCI-propose-tx.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+transactions are then pushed on top of the transactions received from
+`mempool.Reap`.
+
+### `DeliverTx`
+
+Since the list of `tx` received from `ProposeTx` are _not_ passed through `CheckTx`,
```
Comment (milosevic) [inline docs/architecture/adr-012-ABCI-propose-tx.md]: In both cases, `CheckTx` and `ProposeTx`, the application is the one ensuring the correctness of the transactions. So the ProposeTx does not change the model. It might however still be useful to have a tag that differentiate internal transactions from the external one.

**Comment by iantanwx** [comment]
@milosevic, thanks for your input, I’ll amend the proposal in the next couple of days!

File: docs/architecture/adr-012-ABCI-propose-tx.md (hunk 1 above)
Comment (iantanwx) [reply docs/architecture/adr-012-ABCI-propose-tx.md]: Yes, you’re right. It seems the code base has changed slightly since I wrote this proposal.

Comment (milosevic) [reply docs/architecture/adr-012-ABCI-propose-tx.md]: Maybe use of permanent links could help. See this: https://help.github.com/articles/creating-a-permanent-link-to-a-code-snippet/.

**Comment by jaekwon** [comment]
It doesn't make sense to me that Tendermint blocks would be Plasma child chain blocks.  Plasma requires child chains to manage re-orgs on the root proof of work chain... and Tendermint blocks shouldn't depend on external signals for re-orging blocks -- Tendermint blocks should never fork.

The only way I see Plasma Tendermint child-chains working is by creating virtual blocks, and the proposer (perhaps) submitting the latest deposit/exit transactions as pending txs upon which a branching tree of virtual blocks depend on, so that (a) the pending txs can be finalized over time by validators or "witnesses" in peggy verifying root block finality over time (this is what peggy does) and (b) pruning deprecated virtual block branches depending on the finality as determined by (a).

It definitely doesn't make sense to introduce external signals that cause Tendermint blocks to re-org, those signals should be introduced as txs, either as mempool txs or proposer txs are you specified here.  

Maintaining a branching tree of virtual blocks and pruning/finalizing them within the app-state is a bit complicated, but it can be done, and it's much easier for UTXOs.

> This naive approach [JAE: the approach I described just above] is, however, fundamentally flawed, as it by definition diverges from the canonical chain maintained by Tendermint.

I don't think this should necessarily be viewed as flawed -- Tendermint blocks are solving the problem of atomic broadcast across the validator set and full nodes observing the chain, which is arguably an independent concern from the construction of blocks that need to depend on external signals (e.g. what is the latest status of the PoW chain which may get reverted)... and...

> Additionally, this has serious implications for "watchers" - independent third parties, or even an auxilliary blockchain, responsible for ensuring that blocks recorded on the Root Chain are consistent with the Plasma chain's.

... Watchers can still observe the virtual blocks within the Tendermint app's state, keyed the the identifier of the virtual block, probably its hash.  It doesn't seem to be a major problem to me.

> Since, in this case, the Plasma chain is inconsistent with the canonical one maintained by Tendermint Core, it seems that there exists no compact means of verifying the legitimacy of the Plasma chain without replaying every state transition from genesis (!).

Same as above, if the two types of blocks (tendermint & virtual-child-chain) are deemed to be serving different functions, I don't think there is a problem.

TLDR; the real problem is dealing with the forking nature of the root chain, and the way to solve that is with virtual blocks, with or without ProposerTxs.  Even the tracking of root block deposit/exit transactions should be done by mempool txs of validators signing off on their observations, not tied to the block production frequency of the Tendermint chain, like how Peggy would work.  ProposerTxs probably have a good use-case but to the best of my knowledge this is probably not it.  I think the fact that virtual blocks are UTXOs help make virtual block verification and virtual pruning relatively cheap and easy, and observers can definitely observe these virtual blocks, so I think that's the way to go.
